src/Components/WebAssembly/WebAssembly/src/HotReload/HotReloadAgent.cs

```diff
@@ -4,227 +4,131 @@
-            {
-                if (method.Name == name)
-                {
-                    _log($"Type '{handlerType}' has method '{method}' that does not match the required signature.");
-                    break;
-                }
-            }
-
-            return null;
+            Reporter.Report($"Error retrieving capabilities: {e.Message}", AgentMessageSeverity.Error);
         }
     }
 
-    internal static List<Assembly> TopologicalSort(Assembly[] assemblies)
-    {
-        var sortedAssemblies = new List<Assembly>(assemblies.Length);
+    public string Capabilities => _capabilities ?? string.Empty;
 
-        var visited = new HashSet<string>(StringComparer.Ordinal);
+    private void OnAssemblyLoad(object? _, AssemblyLoadEventArgs eventArgs)
+    {
+        _metadataUpdateHandlerInvoker.Clear();
 
-        foreach (var assembly in assemblies)
+        var loadedAssembly = eventArgs.LoadedAssembly;
+        var moduleId = TryGetModuleId(loadedAssembly);
+        if (moduleId is null)
         {
-            Visit(assemblies, assembly, sortedAssemblies, visited);
+            return;
         }
 
-        [UnconditionalSuppressMessage("Trimming", "IL2026", Justification = "Hot reload is only expected to work when trimming is disabled.")]
-        static void Visit(Assembly[] assemblies, Assembly assembly, List<Assembly> sortedAssemblies, HashSet<string> visited)
+        if (_deltas.TryGetValue(moduleId.Value, out var updateDeltas) && _appliedAssemblies.TryAdd(loadedAssembly, loadedAssembly))
         {
-            var assemblyIdentifier = assembly.GetName().Name!;
-            if (!visited.Add(assemblyIdentifier))
-            {
-                return;
-            }
-
-            foreach (var dependencyName in assembly.GetReferencedAssemblies())
-            {
-                var dependency = Array.Find(assemblies, a => a.GetName().Name == dependencyName.Name);
-                if (dependency is not null)
-                {
-                    Visit(assemblies, dependency, sortedAssemblies, visited);
-                }
-            }
-
-            sortedAssemblies.Add(assembly);
+            // A delta for this specific Module exists and we haven't called ApplyUpdate on this instance of Assembly as yet.
+            ApplyDeltas(loadedAssembly, updateDeltas);
         }
-
-        return sortedAssemblies;
     }
 
-    public void ApplyDeltas(IReadOnlyList<UpdateDelta> deltas)
+    public IReadOnlyCollection<(string message, AgentMessageSeverity severity)> GetAndClearLogEntries(ResponseLoggingLevel loggingLevel)
+        => Reporter.GetAndClearLogEntries(loggingLevel);
+
+    public void ApplyDeltas(IEnumerable<UpdateDelta> deltas)
     {
-        for (var i = 0; i < deltas.Count; i++)
+        Debug.Assert(Capabilities.Length > 0);
+        Debug.Assert(_applyUpdate != null);
```

Comment:
Isn't it possible for these assertions to fail if `GetUpdaterMethodsAndCapabilities()` doesn't succeed? It looks like `ApplyDeltas()` gets called before `GetAndClearLogEntries()`, which could mean that this method might throw before the reporter's messages are retrieved and logged (the call to `_applyUpdate` might throw an NRE).

I wonder if we should instead add reporter entries and return early if either of these conditions are met.

I might be missing something about how the flow of everything works and maybe this is actually fine, but my concern is that the `Debug.Assert` doens't represent an invariant.

Reply:
You're right. Let me refactor this a bit.